lib/stripe/gateway.rb

```diff
@@ -0,0 +1,72 @@
+# frozen_string_literal: true
+
+require 'stripe'
+require 'stripe/input_data_validator'
+require 'stripe/result'
+
+module Stripe
+  class Gateway
+    def self.charge(data)
+      new(data).charge
+    end
+
+    def initialize(data)
+      @data = data
+      @api_errors = []
+    end
+
+    def charge
+      Result.new(charge_result, errors)
+    end
+
+    private
+
+    attr_reader :data, :api_errors
+
+    def charge_result
+      charge_and_rescue_data_related_errors
+    rescue Stripe::RateLimitError
+      save_error(:too_many_requests)
+    rescue Stripe::APIConnectionError
+      save_error(:connection_problems)
+    rescue Stripe::StripeError
+      save_error(:stripe_error)
+    rescue StandardError
+      save_error(:unknown_error)
+    end
+
+    def charge_and_rescue_data_related_errors
+      create_stripe_charge
+    rescue Stripe::AuthenticationError
+      save_error(:invalid_api_key)
+    rescue Stripe::InvalidRequestError
+      save_error(:invalid_parameter)
+    rescue Stripe::CardError
+      save_error(:declined_card)
+    rescue Stripe::APIError
+      save_error(:invalid_response_object)
+    end
```

Comment:
This is missing `StandardError`

Reply:
It's here:
https://github.com/survival/donation-system/blob/5745e7dd845ea48bccf7fe40090f5be03799cad0/lib/stripe/gateway.rb#L34

Reply:
Oh, I see the flow now - need more coffee! What's the logic behind splitting the exception handling across two methods?

Also, might this be better?

```
STRIPE_EXCEPTION_MAPPING = {
  Stripe::RateLimitError => : too_many_requests,
  Stripe::APIConnectionError => : :connection_problems
  ...
}.freeze
```

And then the rescue block is simply:

```
  rescue STRIPE_EXCEPTION_MAPPING.keys => e
    save_error(STRIPE_EXCEPTION_MAPPING.class)
```

Just an idea...

Reply:
Looks nice, I'll think about it. In the future, we would want to send the error to `save_error`, so we may not have symbols anymore, and we may have something like:

```
rescue [array of Stripe exceptions] => error
  save_error(error)
rescue Maybe some special exception that we want to handle differently? => error
  save_error(error) or do something else
end
```
and then we won't need the hash, so for now I will not introduce it but I will keep it in mind :+1: 
This could also allow to separate the errors we want to send to the GUI from the errors we want to log for devs, for example, but who knows! :crystal_ball: 

The reason for splitting was rubocop complaining that the method was too long, as well as my own eyes crying blood, so for the time being I grouped exceptions by type, but all this will probably disappear in the future.